src/logging.c

```diff
@@ -134,17 +160,102 @@ void rcutils_logging_console_output_handler(
+    char * token_buffer = static_token_buffer;
+    size_t j;
+    // Look for a token end delimiter.
+    for (j = i + 1; j < size && str[j] != token_end_delimiter; j++) {
+    }
+    if (j >= size) {
+      // No end delimiters found in the remainder of the format string;
+      // there won't be any more tokens so shortcut the rest of the checking
+      // TODO(dhood) before merge: dynamically allocate more space if required.
+      strncat(output_buffer, token_buffer, n);
+      break;
+    }
+    // Found what looks like a token; determine if it's recognized.
+    size_t token_len = j - i - 1;  // not including delimiters
+    strncpy(token, str + i + 1, token_len);
+    if (strcmp("severity", token) == 0) {
+      n = rcutils_snprintf(token_buffer, sizeof(static_token_buffer), "%s", severity_string);
+    } else if (strcmp("name", token) == 0) {
+      n = rcutils_snprintf(token_buffer, sizeof(static_token_buffer), "%s", name);
+    } else if (strcmp("message", token) == 0) {
+      // This has already been expanded above and is able to be larger than the static size.
+      token_buffer = message_buffer;
+      n = strlen(token_buffer);
+    } else if (strcmp("function_name", token) == 0) {
+      n =
+        rcutils_snprintf(token_buffer, sizeof(static_token_buffer), "%s", location->function_name);
+    } else if (strcmp("file_name", token) == 0) {
+      n = rcutils_snprintf(token_buffer, sizeof(static_token_buffer), "%s", location->file_name);
+    } else if (strcmp("line_number", token) == 0) {
+      n = rcutils_snprintf(token_buffer, sizeof(static_token_buffer), "%zu", location->line_number);
+    } else {
+      // This wasn't a token; print the start delimiter and continue the search as usual
+      // (it might contain more start delimiters)
+      // TODO(dhood) before merge: dynamically allocate more space if required.
+      strncat(output_buffer, str + i, 1);
+      continue;
+    }
+
+    if (n >= 0) {
+      size_t required_output_buffer_size = strlen(output_buffer) + n;
+      // printf("Token expansion size=%i, Required output buffer size will be: %zu\n",
+      // n, required_output_buffer_size);
+      if (required_output_buffer_size >= output_buffer_size) {
+        // Token expansion won't fit: allocate more memory dynamically.
+        do {
+          output_buffer_size *= 2;
+        } while (required_output_buffer_size >= output_buffer_size);
+        void * dynamic_output_buffer = allocator.allocate(output_buffer_size, allocator.state);
+        strncpy(dynamic_output_buffer, output_buffer, strlen(output_buffer));
+        // TODO(dhood): uncomment this once it's tested.
+        /*
+        if (output_buffer != static_output_buffer) {
+          allocator.deallocate(output_buffer, allocator.state);
+        }
+        */
+        output_buffer = (char *)dynamic_output_buffer;
+        // printf("New output buffer size: %zu\n", output_buffer_size);
+      }
+      strncat(output_buffer, token_buffer, n);
```

Comment:
Windows will complain about the `strcpy` here (at least it has for me in the past). `memcpy` does the trick.

Reply:
That comment was supposed to go on line 234.

Reply:
thanks for the tip!